# Experiment 5: Fuzzy Set Opeartions
**Name:** Arshan Attar  
**Roll Number:** 231408  


1.   **Classsical set** - A classical set is defined as in which the values are either fully in or fully out , a yes or no .

2.   **Fuzzy set** - A fuzzy set is defined as a set where the members of the set have a degree of membership.


**Operations on Fuzzy set -**

1.  **Union** = It is the maximum of the degree of membership of the two sets
2.  **Intersection** = It is the minimum of the degree of membership of the two sets .
1.  **Complement** = It means negation of the degree of membership i.e 1-degree of membership.
2.  **Substraction** = substraction of two sets a and be is indicated by a-b = a intersection ~b  



1.  **Algebraic Sum** = It is the combination of two fuzzy sets where the membership values are calculated by adding the degree of membership of both sets and subtracting their product.
2.  **Algebraic Product** = It is the multiplication of the degree of membership values of two fuzzy sets.


1.  **Bounded Sum** = It is the addition of the degree of membership values of two fuzzy sets, where the maximum membership value is limited to 1.
2.  **Bounded Product** = It is the subtraction of 1 from the sum of membership values of two fuzzy sets, where negative values are replaced by 0.







Importing the numpy library to perform operaions on the fuzzy set

In [ ]:
import numpy as np


Declaring the sets A and B i.e giving the degree of membership .

In [ ]:

A = np.array([0.2, 0.9, 0.6, 0.4, 0.8])
B = np.array([0.5, 0.4, 0.3, 0.9, 0.7])


Union is the maximum of the membership degree of sets A and B

In [ ]:
union = np.maximum(A, B)
print("Union:", union)

Union: [0.5 0.9 0.6 0.9 0.8]


Intersection if the minimum the membership degree of two sets

In [ ]:
intersection = np.minimum(A, B)
print("Intersection:", intersection)

Intersection: [0.2 0.4 0.3 0.4 0.7]


complement is the negation of a membership degreee of a particular set

In [ ]:
complement = 1 - A
print("Complement:", complement)

Complement: [0.8 0.1 0.4 0.6 0.2]


diffrence is substracting membership degree of the two sets , negative values will not be considered

In [ ]:
C = 1 - B
print(C)
subtraction = np.minimum(A,C)
print("Subtraction:", subtraction)

[0.5 0.6 0.7 0.1 0.3]
Subtraction: [0.2 0.6 0.6 0.1 0.3]


Manual calculation:
*   A = 0.2, 0.9, 0.6, 0.4, 0.8
*   B = 0.5, 0.4, 0.3, 0.9, 0.7





1. Union i.e maximum - max(0.2 & 0.5) = 0.5 ; max(0.9 & 0.4) = 0.9 ; max(0.6 & 0.3) = 0.6 ; max(0.4 & 0.9) = 0.9 ; max(0.8 & 0.7) = 0.8

    Union = [0.5, 0.9, 0.6, 0.9, 0.8]

2. Intersection i.e minimum - min(0.2 & 0.5) = 0.2 ; min(0.9 & 0.4) = 0.4 ; min(0.6 & 0.3) = 0.3 ; min(0.4 & 0.9) = 0.4 ; min(0.8 & 0.7) = 0.7

    Intersection = [0.2, 0.4, 0.3, 0.4, 0.7]

3. Complement i.e negation i.e 1 - A =

    1-0.2, 1-0.9, 1-0.6, 1-0.4, 1-0.8 = 0.8, 0.1, 0.4, 0.6, 0.2

    Complement = [0.8, 0.1, 0.4, 0.6, 0.2]

4. Substraction i.e A-B = A intersection ~B

    B' = 1-B = [1-0.5, 1-0.4, 1-0.3, 1-0.9, 1-0.7] = [0.5, 0.6, 0.7, 0.1, 0.3]

    A-B = min(A, B')

    = [min(0.2,0.5), min(0.9,0.6), min(0.6,0.7), min(0.4,0.1), min(0.8,0.3)]

    = [0.2, 0.6, 0.6, 0.1, 0.3]

5. Algebraic Sum i.e A+B-(A×B)

    = (0.2+0.5)-(0.2×0.5), (0.9+0.4)-(0.9×0.4), (0.6+0.3)-(0.6×0.3), (0.4+0.9)-(0.4×0.9), (0.8+0.7)-(0.8×0.7)

    = 0.7-0.1, 1.3-0.36, 0.9-0.18, 1.3-0.36, 1.5-0.56

    = [0.60, 0.94, 0.72, 0.94, 0.94]

6. Algebraic Product i.e A×B

    = (0.2×0.5), (0.9×0.4), (0.6×0.3), (0.4×0.9), (0.8×0.7)

    = 0.10, 0.36, 0.18, 0.36, 0.56

    Algebraic Product = [0.10, 0.36, 0.18, 0.36, 0.56]

7. Bounded Sum i.e min(1, A+B)

    = min(1,0.2+0.5), min(1,0.9+0.4), min(1,0.6+0.3), min(1,0.4+0.9), min(1,0.8+0.7)

    = min(1,0.7), min(1,1.3), min(1,0.9), min(1,1.3), min(1,1.5)

    = [0.7, 1.0, 0.9, 1.0, 1.0]

8. Bounded Product i.e max(0, A+B-1)

    = max(0,0.2+0.5-1), max(0,0.9+0.4-1), max(0,0.6+0.3-1), max(0,0.4+0.9-1), max(0,0.8+0.7-1)

    = max(0,-0.3), max(0,0.3), max(0,-0.1), max(0,0.3), max(0,0.5)

    = [0, 0.3, 0, 0.3, 0.5]





